File: CPP/ByTier/G/2127_G1_Inter_Active_Easy_Version.cpp
// Problem: CF 2127 G1 - Inter Active (Easy Version)
// https://codeforces.com/contest/2127/problem/G1

/*
 * Problem: Inter Active (Easy Version)
 * Description:
 *   Given a hidden permutation p of length n (where p_i != i for all i),
 *   find the permutation using interactive queries.
 *   Each query provides a permutation q and returns the number of pairs (i,j) such that:
 *     - i < j
 *     - p[q[i]] = q[j]
 *     - i != k (where k is a fixed value chosen at the start)
 *
 * Algorithm:
 *   1. Choose k = (n + 1) / 4 + 1.
 *   2. For each element in the permutation, determine its position using binary search-like technique.
 *   3. Use queries to narrow down candidates and deduce the permutation.
 *
 * Time Complexity: O(n^2 * log n) - Each element needs up to O(n) queries, and there are up to O(n) elements,
 *                                   and logarithmic number of steps for binary search.
 *
 * Space Complexity: O(n) - To store arrays and temporary vectors of size n.
 */

#include <algorithm>
#include <iostream>
#include <stdlib.h>
#include <utility>
#include <vector>
using namespace std;
const int MXN = 1005;
int n, k;
int p[MXN], qarr[MXN];
bool used[MXN];

// Ask the jury for a query result
int ask() {
  cout << "? ";
  for (int i = 1; i <= n; i++)
    cout << qarr[i] << ' ';
  cout << endl;
  int r;
  if (!(cin >> r))
    exit(0);
  if (r == -1)
    exit(0);
  return r;
}

// Output the final permutation
void answer() {
  cout << "! ";
  for (int i = 1; i <= n; i++)
    cout << p[i] << ' ';
  cout << endl;
}

// Construct a query array for a specific target value
void build_query(int fix, vector<int> good, vector<int> bad) {
  fill(used + 1, used + n + 1, false);
  fill(qarr + 1, qarr + n + 1, 0);
  qarr[k] = fix;
  used[fix] = true;
  for (int i = 1; i < k && !good.empty(); i++) {
    qarr[i] = good.back();
    used[qarr[i]] = true;
    good.pop_back();
  }
  for (int i = n; !good.empty(); i--) {
    qarr[i] = good.back();
    used[qarr[i]] = true;
    good.pop_back();
  }
  for (int i = k + 1; !bad.empty(); i++) {
    qarr[i] = bad.back();
    used[qarr[i]] = true;
    bad.pop_back();
  }
  vector<int> rest;
  for (int v = 1; v <= n; v++)
    if (!used[v])
      rest.push_back(v);
  for (int i = 1; i <= n; i++)
    if (qarr[i] == 0) {
      qarr[i] = rest.back();
      rest.pop_back();
    }
}

// Reverse the elements except the element at position k
void rev_except_k() {
  vector<int> v;
  v.reserve(n - 1);
  for (int i = 1; i <= n; i++)
    if (i != k)
      v.push_back(qarr[i]);
  for (int i = 1, j = (int)v.size() - 1; i <= n; i++)
    if (i != k) {
      qarr[i] = v[j--];
    }
}

// Solve one test case
void solve_one() {
  cin >> n;
  k = (n + 1) / 4 + 1;
  cout << k << endl;
  for (int target = 1; target <= n; target++) {
    vector<int> cand;
    cand.reserve(n - 1);
    for (int v = 1; v <= n; v++)
      if (v != target)
        cand.push_back(v);
    while ((int)cand.size() > 1) {
      int capB = n - (2 * k - 1); // Maximum bad elements allowed
      int takeBad = min((int)((cand.size() + 1) / 2), capB); // Number of elements to place in "bad" part
      int keepGood = (int)cand.size() - takeBad; // Remaining elements for "good"
      vector<int> good(cand.begin(), cand.begin() + keepGood);
      vector<int> bad(cand.begin() + keepGood, cand.end());
      build_query(target, good, bad); // Construct a query with good and bad parts
      int a = ask();
      rev_except_k(); // Reverse except k and ask again
      int b = ask();
      if (a + b == n - 1) {
        cand = move(good);
      } else {
        cand = move(bad);
      }
    }
    int pred = cand.back();
    p[pred] = target;
  }
  answer();
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int T;
  if (!(cin >> T))
    return 0;
  while (T--)
    solve_one();
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Java/ByTier/G/2127_G1_Inter_Active_Easy_Version.java
// Problem: CF 2127 G1 - Inter Active (Easy Version)
// https://codeforces.com/contest/2127/problem/G1

/*
 * Problem: Inter Active (Easy Version)
 * Purpose: Find a hidden permutation p of length n using at most 15*n queries.
 * Algorithm:
 *   - Select k = (n+1)/4 + 1 as a constant.
 *   - For each position i in the result:
 *     - Build a query permutation q where q[k] = i.
 *     - Partition remaining positions into left and right parts.
 *     - Query using this partition and use the response to narrow down the search.
 *     - Repeat until the correct position is found.
 *   - Time Complexity: O(n^2 * log n)
 *   - Space Complexity: O(n)
 */

import java.io.*;
import java.util.*;

public class Main {
    static final int MXN = 1003;
    static int n, k;
    static int[] p = new int[MXN];
    static int[] q = new int[MXN];
    static boolean[] mark = new boolean[MXN];
    static BufferedReader br;
    static PrintWriter out;

    // Ask the jury for the number of valid pairs
    static int ask() throws IOException {
        out.print("? ");
        for (int i = 1; i <= n; i++) {
            out.print(q[i]);
            if (i < n) out.print(' ');
        }
        out.println();
        out.flush();
        String line = br.readLine();
        if (line == null) System.exit(0);
        line = line.trim();
        if (line.equals("-1")) System.exit(0); 
        return Integer.parseInt(line);
    }

    // Output the final permutation
    static void ans() {
        out.print("! ");
        for (int i = 1; i <= n; i++) {
            if (i > 1) out.print(' ');
            out.print(p[i]);
        }
        out.println();
        out.flush();
    }

    // Build query permutation q for a given value at position k
    static void build(int valI, ArrayList<Integer> opt, ArrayList<Integer> optbad) {
        Arrays.fill(mark, false);
        Arrays.fill(q, 0);
        q[k] = valI;
        mark[valI] = true;
        
        // Assign elements to positions [1..k-1]
        for (int j = 1; j < k && !opt.isEmpty(); j++) {
            int x = opt.remove(opt.size() - 1);
            q[j] = x;
            mark[x] = true;
        }

        // Assign remaining elements in reverse manner to [n down to k+1]
        for (int j = n; !opt.isEmpty(); j--) {
            int x = opt.remove(opt.size() - 1);
            q[j] = x;
            mark[x] = true;
        }

        // Assign elements from bad list
        for (int j = k + 1; !optbad.isEmpty(); j++) {
            int x = optbad.remove(optbad.size() - 1);
            q[j] = x;
            mark[x] = true;
        }

        // Fill remaining zeros with unused elements
        ArrayList<Integer> remaining = new ArrayList<>();
        for (int j = 1; j <= n; j++) {
            if (!mark[j]) remaining.add(j);
        }
        for (int j = 1; j <= n; j++) {
            if (q[j] == 0) {
                q[j] = remaining.remove(remaining.size() - 1);
            }
        }
    }

    // Reverse query array
    static void rev() {
        ArrayList<Integer> vec = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (i != k) vec.add(q[i]);
        }
        for (int i = 1; i <= n; i++) {
            if (i == k) continue;
            q[i] = vec.remove(vec.size() - 1);
        }
    }

    static void solveCase() throws IOException {
        n = Integer.parseInt(br.readLine().trim());
        k = (n + 1) / 4 + 1;
        out.println(k);
        out.flush();

        // For each index i, find where it should be placed in the permutation
        for (int i = 1; i <= n; i++) {
            ArrayList<Integer> opt = new ArrayList<>();
            for (int j = 1; j <= n; j++) {
                if (j != i) opt.add(j);
            }

            // Binary search to determine correct element at position i
            while (opt.size() > 1) {
                int sz = opt.size() - Math.min((opt.size() + 1) / 2, n - (2 * k - 1));
                if (sz < 0) sz = 0;
                if (sz > opt.size()) sz = opt.size();

                ArrayList<Integer> left = new ArrayList<>(opt.subList(0, sz));
                ArrayList<Integer> right = new ArrayList<>(opt.subList(sz, opt.size()));

                build(i, new ArrayList<>(left), new ArrayList<>(right));
                int val = ask();
                
                rev(); // Reverse the query and make another call
                val += ask();

                if (val == n - 1) {
                    opt = left;
                } else {
                    opt = right;
                }
            }
            
            // Once the element is determined, map it to its correct position
            int pos = opt.get(0);
            p[pos] = i;
        }
        ans();
    }

    public static void main(String[] args) throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        String first = br.readLine();
        if (first == null) return;
        first = first.trim();
        int tc;
        try {
            tc = Integer.parseInt(first);
        } catch (NumberFormatException e) {
            return;
        }
        for (int t = 0; t < tc; t++) {
            solveCase();
        }
        out.close();
    }
}


// https://github.com/VaHiX/CodeForces/